File: Java/DeepSpace/src/deepspace/Dice.java
package deepspace;
import java.util.Random;
/**
 *
 * @author victor
 */
class Dice {
    private final float NHANGARSPROB ,NSHIELDSPROB, NWEAPONSPROB, FIRSTSHOTPROB;
    private Random generator;

    public Dice() {
        NHANGARSPROB=0.25f;
        NSHIELDSPROB=0.25f;
        NWEAPONSPROB=0.33f;
        FIRSTSHOTPROB=0.5f;
        generator = new Random();
    }
    
    int initWithHangars() {
       if(generator.nextFloat() < NHANGARSPROB)
           return 0;
       return 1;
    }
    
    int initWithNWeapons() {
        float aux = generator.nextFloat();
        if(aux < NWEAPONSPROB)
            return 1;
        else if (aux < NWEAPONSPROB*2)
            return 2;
        return 3;
    }
        
    int initWithNShields() {
       if(generator.nextFloat() < NSHIELDSPROB)
           return 0;
       return 1;
    }
    
    int whoStarts(int nPlayers) {
        return generator.nextInt(nPlayers);
    }
    
    GameCharacter firstShot() {
        if(generator.nextFloat() > FIRSTSHOTPROB)
           return GameCharacter.ENEMYSTARSHIP;
        else
            return GameCharacter.SPACESTATION;
    }
    
    boolean spaceStationMoves(float speed) {
        return generator.nextFloat() <= speed;
    }
    
    
    
    
    
}

File: Deepspace_ruby/Deepspace/lib/Dice.rb
#encoding:utf-8
require_relative "GameCharacter"

module Deepspace

class Dice   
  def initialize
    @generator = Random.new
    @NHANGARSPROB = 0.25
    @NSHIELDSPROB = 0.25
    @NWEAPONSPROB = 0.33
    @FIRSTSHOTPROB = 0.5
    
  end
  
  def initWithNHangars
    if @generator.rand(1.0) >= @NHANGARSPROB
      return 1
    else
      return 0
    end
  end
  
  def initWithNWeapons
    aux = @generator.rand(1.0)
    if aux < @NWEAPONSPROB
      return 1
    elsif aux < 2*@NWEAPONSPROB
      return 2
    else 
      return 3
    end
  end
  
  def initWithNShields
    if @generator.rand(1.0) <= @NSHIELDSPROB
      return 0
    else
       return 1
    end
  end
  
  def whoStarts(nPlayers)
    return @generator.rand(nPlayers) #Con ... excluye el límite superior, con ..
    #lo incluye  
  end
  
  def firstShot
     if @generator.rand(1.0) <= @FIRSTSHOTPROB
      return GameCharacter::SPACESTATION
    else
       return GameCharacter::ENEMYSTARSHIP
    end
  end
  
  def spaceStationMoves(speed)
    @generator.rand(1.0) <= speed
  end
  
  def to_s
       return "\nNHANGARSPROB = #{@NHANGARSPROB} NSHIELDSPROB = #{@NSHIELDSPROB} NWEAPONSPROB = #{@NWEAPONSPROB} FIRSTSHOTPROB = #{@FIRSTSHOTPROB}"
end
end
end
